crates/compiler/gen_wasm/src/low_level.rs

```diff
@@ -2117,6 +2117,17 @@ pub fn call_higher_order_lowlevel<'a>(
         ..
     } = passed_function;
 
+    // The zig lowlevel builtins expect the passed functions' closure data to always
+    // be sent as an opaque pointer. On the Roc side, however, we need to call the passed function
+    // with the Roc representation of the closure data. There are three possible cases for that
+    // representation:
+    //
+    // 1. The closure data is a struct
+    // 2. The closure data is an unwrapped value
+    // 3. There is no closure data
+    //
+    // To uniformly deal with the first two cases, always box these layouts before calling the
+    // builtin; the wrapper around the passed function will unbox them.
```

Comment:
What jumps out at me most here is the fact that it's heap-allocated. I'm trying to figure out why that's needed. I think it would be better to use `Layout::Struct` here rather than `Layout::Box`, if we can get away with it.

In case (1), the closure data is a struct, whose size is known at compile time
In case (2), the closure data is an unwrapped value... whose size is also known at compile time, I think? Or am I misunderstanding what you mean by "unwrapped value"? I'm assuming it's an integer?

Reply:
It is known at compile time, but the problem is that it might not be an `i32`. The particular test that induced this change passes a captured `f64`. We can't just pass the `f64` as the closure data to zig, because then we'll get a type error.

Reply:
Ah OK. But it does sound like we can use a struct instead of a box though. That'll still be an `i32` pointer, just pointing to stack memory instead of heap memory. That should be faster.
So to do that you'll have to add a new variable, as you have done, but the layout for that variable will be a struct containing the `f64` or whatever. Stack space will be created for it.

There's a helper function at the bottom of this file called `fn ensure_symbol_is_in_memory` and it is doing pretty much the same as what you need to do here. It the symbol is a number, it stores it in a struct in stack memory. If it was already in stack memory it doesn't bother. Either way, it gives you back the info you need to load the pointer.